config_ext_test.go

```diff
@@ -0,0 +1,350 @@
+
+	t.Run("multiple gets", func(t *testing.T) {
+		var s string
+		run(TestCase{
+			Value:     provider.Get("occupants").Get("honda").Get("driver"),
+			HasValue:  true,
+			Interface: "jane",
+			Populate:  &s,
+		})
+		assert.Equal(t, "jane", s, "unexpected result after populate")
+	})
+
+	t.Run("default missing value", func(t *testing.T) {
+		var s string
+		run(TestCase{
+			Value:     provider.Get("occupants").Get("honda").Get("driver"),
+			HasValue:  true,
+			Interface: "jane",
+			Populate:  &s,
+		})
+		assert.Equal(t, "jane", s, "unexpected result after populate")
+	})
+
+	t.Run("default overriden by scalar", func(t *testing.T) {
+		var s string
+		val, err := provider.Get("practical.honda").WithDefault("CRV")
+		require.NoError(t, err, "couldn't set default")
+		run(TestCase{
+			Value:     val,
+			HasValue:  true,
+			Interface: "civic",
+			Populate:  &s,
+		})
+		assert.Equal(t, "civic", s, "unexpected result after populate")
+	})
+
+	t.Run("default overriden by nil", func(t *testing.T) {
+		t.Skip("TODO: panics")
+		var s string
+		val, err := provider.Get("nothing").WithDefault("something")
+		require.NoError(t, err, "couldn't set default")
+		run(TestCase{
+			Value:     val,
+			HasValue:  true,
+			Interface: nil,
+			Populate:  &s,
+		})
+		assert.Equal(t, "", s, "unexpected result after populate")
+	})
+
+	t.Run("default merges maps", func(t *testing.T) {
+		var m map[string]string
+		val, err := provider.Get("practical").WithDefault(map[string]string{"ford": "fiesta"})
+		require.NoError(t, err, "couldn't set default")
+		run(TestCase{
+			Value:    val,
+			HasValue: true,
+			Interface: map[interface{}]interface{}{
+				"honda":  "civic",
+				"toyota": "camry",
+				"nissan": "altima",
+				"ford":   "fiesta",
+			},
+			Populate: &m,
+		})
+		assert.Equal(t, map[string]string{
+			"honda":  "civic",
+			"toyota": "camry",
+			"nissan": "altima",
+			"ford":   "fiesta",
+		}, m, "unexpected result after populate")
+	})
+
+	t.Run("default replaces sequences", func(t *testing.T) {
+		var s []string
+		val, err := provider.Get("fun").WithDefault([]string{"delorean"})
+		require.NoError(t, err, "couldn't set default")
+		run(TestCase{
+			Value:     val,
+			HasValue:  true,
+			Interface: []interface{}{"maserati", "lamborghini"},
+			Populate:  &s,
+		})
+		assert.Equal(t, []string{"maserati", "lamborghini"}, s, "unexpected result after populate")
+	})
+
+	t.Run("default ignores type mismatches", func(t *testing.T) {
+		// TODO: We should return an error in strict mode.
+		// fun is a sequence, so we shouldn't be able to merge in a mapping
+		val, err := provider.Get("fun").WithDefault(map[string]string{"dodge": "viper"})
+		assert.NoError(t, err, "error on type mismatch in permissive mode")
+		assert.Equal(t, []interface{}{"maserati", "lamborghini"}, val.Value(), "wrong merged output")
+	})
+
+	t.Run("chained defaults", func(t *testing.T) {
+		// Defaults should be FIFO: once we've set a default value, we can't later
```

Comment:
nit: not sure if FIFO is the best description here since that might imply the next-queued default would be used on a subsequent call to `Get` (i.e. `boom` would be used after `bottom`). Seems strange to use "idempotent" in this context, but something similar to "no effect after the first call", might be more appropriate.

Reply:
Also a good point, I'll reword this.